### PR TITLE
feat(payment_attempt): add kv for find last successful attempt

### PR DESCRIPTION
**Pull request by dracarys18**
## Type of Change

- [x] Enhancement

## Description

Add KV for last successful payment attempt

## Motivation and Context

Add it in KV for better latency

## How did you test it?

This function is being used in `refunds_create_core`. So refund should succeed.

Screenshot 2023-09-20 at 3 06 54 PM

Screenshot 2023-09-20 at 3 07 15 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code

### CODE REVIEW

File: crates/storage_impl/src/payments/payment_attempt.rs [hunk 1]
```diff
@@ -373,6 +373,23 @@ impl<T: DatabaseStore> PaymentAttemptInterface for KVRouterStore<T> {
...
                                 &created_attempt.merchant_id, &created_attempt.attempt_id,
                             ),
+                            pk_id: key.clone(),
+                            sk_id: field.clone(),
+                            source: "payment_attempt".to_string(),
+                        }
+                        .insert(&conn)
+                        .await
+                        .map_err(|er| {
+                            let new_err = crate::diesel_error_to_data_error(er.current_context());
```
Comment (jarnura) [inline crates/storage_impl/src/payments/payment_attempt.rs:383]:
```suggestion
                            let new_err = diesel_error_to_data_error(er.current_context());
```

File: crates/storage_impl/src/lookup.rs [hunk 2]
```diff
@@ -50,7 +50,7 @@ impl<T: DatabaseStore> ReverseLookupInterface for RouterStore<T> {
                     er.change_context(new_err)
                 })
         };
-        get_or_populate_redis(self, id, database_call).await
+        get_or_populate_redis(self, format!("reverse_lookup_{id}"), database_call).await
```
Comment (jarnura) [inline crates/storage_impl/src/lookup.rs:53]: In all reverse lookup are we adding prefix as `reverse_lookup_`?

File: crates/storage_impl/src/payments/payment_attempt.rs [hunk 3]
```diff
@@ -570,13 +587,43 @@ impl<T: DatabaseStore> PaymentAttemptInterface for KVRouterStore<T> {
...
+                .find_payment_attempt_last_successful_attempt_by_payment_id_merchant_id(
+                    payment_id,
+                    merchant_id,
+                    storage_scheme,
+                )
+        };
+        match storage_scheme {
+            MerchantStorageScheme::PostgresOnly => database_call().await,
+            MerchantStorageScheme::RedisKv => {
+                let lookup_id = format!("{merchant_id}_{payment_id}");
```
Comment (jarnura) [inline crates/storage_impl/src/payments/payment_attempt.rs:601]: rename `lookup_id` to `partition_key` or just key like other places.

File: crates/storage_impl/src/payments/payment_attempt.rs [hunk 4]
```diff
@@ -570,13 +587,43 @@ impl<T: DatabaseStore> PaymentAttemptInterface for KVRouterStore<T> {
...
+                let lookup = self.get_lookup_by_lookup_id(&lookup_id).await?;
```
Comment (jarnura) [inline crates/storage_impl/src/payments/payment_attempt.rs:602]: why we needed this lookup?

Comment (jarnura) [reply crates/storage_impl/src/payments/payment_attempt.rs:602]: you can construct the lookup key just as `pa_*`.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/storage_impl/src/lookup.rs (hunk 2 above)
Comment (dracarys18) [reply crates/storage_impl/src/lookup.rs:53]: Yeah because this doesn't have shard key so it needs a prefix to avoid key clashing.

**Review by jarnura — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**